### ISSUE TITLE
Unable to implement Hierarchical Persistence as referenced in docs

### ISSUE THREAD

**Issue by jrosmith** [corda/corda#5254]
[Simple Reference Project](https://github.com/jrosmith/corda-persistence)
Documentation Version: 4.0
Corda Version: 4.0

The core of the issue is as follows:

In the docs a child token is tied to the parent token via:
  
   ```
    @Entity
    @CordaSerializable
    @Table(name = "child_data")
    class PersistentChildToken(
            @Id
            var Id: UUID = UUID.randomUUID(),

            // code for other columns

            @ManyToOne(targetEntity = PersistentParentToken::class)
            var persistentParentToken: TokenState

    ) : PersistentState()
```

but whenever i annotate a child with `@Id` i receive the following error:
```
net.corda.nodeapi.internal.persistence.HibernateConfigException: Could not create Hibernate configuration: net.corda.core.schemas.PersistentStateRef must not have @Id properties when used as an @EmbeddedId
```

i've looked through the source code and saw that a PersistentState is defined as the following:
```
@KeepForDJVM
@MappedSuperclass
@CordaSerializable
class PersistentState(@EmbeddedId override var stateRef: PersistentStateRef? = null) : DirectStatePersistable
```

so i guess i understand where the error is coming from, `PersistentChildToken` implements `PersistentState` which already has `@EmbeddedId`. but if i remove the `@Id` field i get the following error:
```
javax.persistence.PersistenceException: org.hibernate.id.IdentifierGenerationException: null id generated for:class
``` 

it seems like i can't use the `@Id` annotation because `PersistentState` is already annotated with `EmbeddedId`, but I also need the `@Id` annotation or else an id will not be properly generated. 

there are further errors i have run into while trying to implement the hierarchal relationship which were eventually solved by adding a `@OneToMany` annotation above `@JoinColumns`. In the context of the docs it would be:
```
@OneToMany(
  cascade = [CascadeType.ALL],
  targetEntity = PersistentChildToken::class
)
```

the project i am attempting implement the hierarchical relationships is more complex than the [simple reference project](https://github.com/jrosmith/corda-persistence), but the core of the workflows, table creation, and relationship mapping is there. 

any help implementing a one to many relationship would be greatly appreciated, thank you.

**Comment by tzdesing** [comment]
Same problem here.. :(

**Comment by prateektiwari7** [comment]
Hi @jrosmith 

Have you solved this issue ?

**Comment by amolpednekar** [comment]
Bumping this! @jrosmith @tzdesing @prateektiwari7 - Were you guys able to resolve this?

PS: This is required to allow Corda's VaultCustomQueryCriteria on the child entities.

**Comment by prateektiwari7** [comment]
Hey @amolpednekar 

Nope it is not solved, please provide us  the way to solve it

**Comment by ivanterziev-r3** [comment]
Marked as a documentation update requirement (for internal tracking see DOC-1812).

**Comment by rach-id** [comment]
@ivanterziev-r3 Can you please provide a brief description on how to solve this issue ? until the documentation is out ?

**Comment by jrosmith** [comment]
@amolpednekar @prateektiwari7 we were not able to resolve this using the hierarchical implementation, but did implement a workaround that requires us to write all queries making use of hierarchical relations as native sql. this has been a frustrating experience but at least something can be hacked together.

For a one to many scenario:
```
// Parent schema implemented as usual
class PersistentParent(
  @Column(name = "linear_id")
  var linearId: UUID
) : PersistentState() {
  contructor() : this( UUID.randomUUID() )
}

// Child has a reference to its parent
class PersistentChild(
  @Column(name = "linear_id")
  var linearId: UUID
  @Column=(name = "parent_linear_id")
  var parentLinearId: UUID
) : PersistentState() {
  contructor() : this( UUID.randomUUID(), UUID.randomUUID() )
}

```

For many-to-many we had to manually create a through state and schema:
```
// Parent schema implemented as usual
class PersistentParent(
  @Column(name = "linear_id")
  var linearId: UUID
) : PersistentState() {
  constructor() : this( UUID.randomUUID() )
}

// Child schema implemented as usual
class PersistentChild(
  @Column(name = "linear_id")
  var linearId: UUID
) : PersistentState() {
  constructor() : this( UUID.randomUUID())
}

// ParentChildThrough table schema
class PersistentThroughTable(
  @Column(name = "parent_linear_id")
  var parentLinearId: UUID
  @Column(name="child_linear_id")
  var childLinearId: UUID
) : PersistentState() {
  constructor() : this( UUID.randomUUID(), UUID.randomUUID() )
}
```

**Comment by tzdesing** [comment]
@amolpednekar, I have used the same approach of @jrosmith, pass UUID in constructor, my problem was in CordaSerializable.. You can Override method too.. If you share the code, we can analyze better..

**Comment by tzdesing** [comment]
@arthurmsouza

**Comment by amolpednekar** [comment]
@jrosmith Your suggested approach requires implementing child entities as Corda states, which I kind of ended up with - but is not ideal and is an additional overhead of maintaining and evolving separately. Especially if the child tables are simple and dont have a lifecyle of their own.

**Comment by amolpednekar** [comment]
The other option like you said is writing native queries on the child table (since it does not extend PersistentState). Because of the up-sert behaviour of custom schemas in Corda, there will be multiple entries per state update, have to put in an additional timestamp and sort by latest to return the latest result.
I was hoping I could rely on Corda's querying APIs to do all this.

**Comment by rach-id** [comment]
@jrosmith Could you please share an example of the manual hierarchical vault queries ?

**Comment by jrosmith** [comment]
@amolpednekar you are absolutely correct about the added overhead and i share your frustration with not being able to use Corda's built in querying apis to handle the situation. unfortunately this approach was the best option available given the circumstances of our use case. if anyone else has found another solution to make this work using the querying api's, i would be interested in hearing it

@SweeXordious some pseudo-code to get you started - note that this logic does not handle consumed/unconsumed states
```
 fun getParentAndChildren(parentLinearId: UUID)  {
   val nativeQuery = getNativeQuery(linearId)
   val session= serviceHub.jdbcSession()
   val rs = session.prepareStatement(nativeQuery).executeQuery()
   
   while (rs.next()) {
     // business logic
    }
 }
 
 fun getNativeQuery(linearId: UUID) : String {
    return """
      select 
        *
      from
        parent_table as p
      join
        child_table as c
      on
        c.parent_linear_id = '$linear_id'::uuid
      where
         p.linear_id = '$linear_id'::uuid
    """.trimIndent()
 }
```

**Comment by rach-id** [comment]
@jrosmith Thank you sooo much for your help

**Comment by nargas-ritu** [comment]
Internal ticket - [CORDA-4212](https://r3-cev.atlassian.net/browse/CORDA-4212)
